### PR TITLE
Create format.yml

### LINKED ISSUE #28: Linter and formatting enforcement

### PR DESCRIPTION
**Pull request by JakeRoggenbuck**
fixes #28

### CODE REVIEW

File: core/include/jml/math/vector.hpp [hunk 1]
```diff
@@ -18,26 +18,25 @@ class JML_API Vector {
...
+    void add(Vector &v);
 
     int get_size() const;
 
     double get_entry(int pos) const;
     void set_entry(int pos, double val);
-    void add_entry(int pos, double val); // Adds val to the current value in pos.
-
+    void add_entry(int pos,
+                   double val); // Adds val to the current value in pos.
```
Comment (adamhutchings) [inline core/include/jml/math/vector.hpp:39]: Any idea where this line break came from?

**Review by adamhutchings — CHANGES_REQUESTED** [review]
I don't know if this is possible, but is there a way to set it so that it enforces `type& variable` instead of `type &variable`? It seems the former is already fairly consistent in our codebase and this would change it.

File: core/include/jml/math/vector.hpp (hunk 1 above)
Comment (JakeRoggenbuck) [reply core/include/jml/math/vector.hpp:39]: I think from the comment. Maybe the comment should be above the line?

**Comment by JakeRoggenbuck** [comment]
> I don't know if this is possible, but is there a way to set it so that it enforces `type& variable` instead of `type &variable`? It seems the former is already fairly consistent in our codebase and this would change it.

I can take a look

File: core/include/jml/math/vector.hpp (hunk 1 above)
Comment (adamhutchings) [reply core/include/jml/math/vector.hpp:39]: Yeah, probably. I think this is fine for now, unless you think otherwise

**Review by adamhutchings — DISMISSED** [review]

File: .clang-format
Comment (Sophon96) [inline .clang-format:1]: I'm gonna guess that this is `BasedOn: LLVM` which is pretty sane.

File: .github/workflows/format.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  build:
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v3
+    - name: format
+      run: |
+        cd core
+        find . -iname "*.hpp" -o -iname "*.cpp" -o -iname "*.c" -o -iname "*.h" | xargs -I {} clang-format -i --dry-run --Werror -style='{IndentWidth: 4, ReferenceAlignment: Left}' {}
```
Comment (Sophon96) [inline .github/workflows/format.yml:19]: We also have one `.tpp` file (template implementation file) :).

Is there a reason why `-i` is passed when `--dry-run` is also passed?

The manual style flag should also be removed.

File: README.md [hunk 3]
```diff
@@ -29,3 +29,14 @@ meson test -C build     # replace `build`
...
 
 Unfortunately, meson doesn't provide good output when a test fails. Running the test executable manually is often more helpful. Look under `build/tests/` to find the right executable. The executable for testing the core library is `core_tests`.
+
+## Format
+
+We use [clang-format](https://www.kernel.org/doc/html/latest/process/clang-format.html) to check our formatting.
+Before you push your code, you can run the following that finds all source code files, and then formats them all in place.
+
+```
+find . -iname "*.hpp" -o -iname "*.cpp" -o -iname "*.c" -o -iname "*.h" | xargs -I {} clang-format -i -style='{IndentWidth: 4, ReferenceAlignment: Left}' {}
```
Comment (Sophon96) [inline README.md]: note about `.tpp` suffix here to

File: core/capi/cjml.cpp [hunk 4]
```diff
@@ -30,14 +30,14 @@ int jml_vector_add_entry(unsigned vid, int pos, double val) {
...
-        case JACT_SIG:
-            a = new jml::Sigmoid();
-        case JACT_RELU:
-            a = new jml::ReLU();
+    case JACT_FSIG:
+        a = new jml::FastSigmoid();
+    case JACT_SIG:
+        a = new jml::Sigmoid();
+    case JACT_RELU:
+        a = new jml::ReLU();
```
Comment (Sophon96) [inline core/capi/cjml.cpp:40]: Not a fan of this personally.

**Review by Sophon96 — COMMENTED** [review]
for some reason, the .clang-format errors on parsing on my machine.

File: core/capi/cjml.cpp (hunk 4 above)
Comment (adamhutchings) [reply core/capi/cjml.cpp:40]: What would you rather it be?

**Comment by adamhutchings** [comment]
@Sophon96 @JakeRoggenbuck what remains?

**Comment by JakeRoggenbuck** [comment]
I fixed the reference align left, clang-format config, and the unneeded flag in the command to run the workflow formatting

**Comment by JakeRoggenbuck** [comment]
> I fixed the reference align left, clang-format config, and the unneeded flag in the command to run the workflow formatting

Anything else @adamhutchings @Sophon96 ?

File: .github/workflows/format.yml (hunk 2 above)
Comment (Sophon96) [reply .github/workflows/format.yml:19]: This needs to be resolved

File: README.md (hunk 3 above)
Comment (Sophon96) [reply README.md]: The `-i` should be retained here but not in CI

**Comment by JakeRoggenbuck** [comment]
That broke something. I will reset to e944bd3

**Comment by JakeRoggenbuck** [comment]
Force pushed the two edits to format check out

**Comment by JakeRoggenbuck** [comment]
Could I please get a review for this?

**Review by adamhutchings — APPROVED** [review]

**Pull request merged**